_posts/2019-04-24-server-rendering-responsively.md

```diff
@@ -0,0 +1,199 @@
+---
+layout: epic
+title: "Server-Rendering A Responsive Artsy.Net"
+date: "2019-04-22"
+author: [steve-hicks, eloy]
+categories: [react, html, web]
+---
+
+At Artsy, we use server-side rendering (SSR) to deliver every page you hit on [artsy.net](https://artsy.net). We
+decided on using SSR for many reasons, amongst them performance. (TODO: Maybe more about why we chose SSR, but I
+don't know much about that history.)
+
+(TODO: Maybe a bit more about SSR & `hydrate`, or just link to those docs.)
+
+One challenge we've run into with SSR is rendering markup and styles from the server that the browser doesn't have
+to throw away. When your browser finishes loading our JavaScript & hydrates our React app, our app suddenly has
+more information about your browser dimensions than we had in the server request. React will go through the
+reconciliation process with this new information, to determine which DOM elements need to be updated. If there are
+significant differences between the markup our server gave you and the component tree our React app is now
+rendering, there's noticeable churn in the DOM.
+
+We'd like our app to churn as little as possible during hydration. We've taken several steps (and mis-steps) toward
+accomplishing this goal: how can we emit _usable_ responsive markup from the server, so that the browser doesn't
+have to throw everything away when the browser dimensions become known?
+
+To help us solve this problem, we've released
+[@artsy/react-responsive-media](https://github.com/artsy/react-responsive-media). This is the story of how
+@artsy/react-responsive-media came about.
+
+<!-- more -->
+
+## Iteration 1: Render props and a disposable component tree
+
+Our initial attempt to emit responsive content from the server is what prompted us to dig deeper into this problem.
+We started with a `<Responsive>` component that used
+[a "render props"/"function as a child" pattern](https://reactjs.org/docs/render-props.html#using-props-other-than-render),
+and allowed you to return different JSX based on the current breakpoint. In practice, it looked something like
+this:
+
+```
+<Responsive>
+  {({ xs }) => {
+    if (xs) {
+      return <Box width={1} />
+    } else {
+      return <Box width={2} />
+    }
+  }}
+</Responsive>
+```
+
+The argument to the child function contained a flag for each of the breakpoints our app supports. In the example
+above, we're rendering a `<Box>` with a width of `1` if the screen matches our `xs` breakpoint; otherwise, we're
+rendering a `<Box>` with a width of `2`.
+
+Our biggest problem with this API was that on the server, we don't know which breakpoint you're at. We had to
+default to something, so we assumed that if we didn't know which breakpoint you wanted, we would give you the
+`xs`/"extra small" one. But if you were browsing with a desktop, the first client-side render will completely throw
```

Comment:
Conventionally this is handled via user-agent detection as described in [react-responsive](https://github.com/ReactTraining/react-media#server-side-rendering-ssr).